enhancements/oc/image-content-source-policy-awareness.md

```diff
@@ -0,0 +1,109 @@
+        if disconnected or not authorized to access image-reference registry
+    * FLAGS:
+        * *--icsp-to* will define where to write an ICSP file to.  If unset, `oc adm release mirror` will write to current directory.
+        * *--icsp-file* will define where to get ICSP from a file.  If set, `oc adm release extract|mirror` will use this ICSP data.
+    * FLAGS DECIDED AGAINST:
+        * boolean `--use-icsp` and if true, check for cluster and/or icsp file?  This is problematic, because even if a user is currently connected to a cluster, 
+        it doesn't mean they want to use information from that cluster with an `oc adm release ...` command.  
+        So here, we'd need a different `--cluster-icsp` boolean and `--icsp-file` string flag.  The flags are adding up here, and that is not desireable. 
+        * string `--image-content-source aregistry/arepo/arelease`. I don't like this, because it would be redundant for a user to run something like this: 
+        `oc adm release extract --command oc myreg:5000/myrepo/release:tag --image-content-source myreg:5000/myrepo/release`.  
+        * Instead, a boolean `--set-prefix` would allow a user to specify "I want to use the prefix of the release image I have specified, 
+        rather than any underlying image reference."
+* This proposal is for the `FLAGS`, that `oc` will try ICSP, user-given image, then fall back on underlying image-reference from user-given image. 
+* See `User Stories` below for examples. 
+
+## Summary
+
+ICSP allows OpenShift (CVO, CRI-O) to check down a list of possible mirrors to find an image with the matching digest it is
+looking for.  `oc` should do the same.  If no ICSP found, then try user-given image.  If that doesn't succeed, then use underlying
+image-reference from user-given image.
+
+## Proposal
+
+* `oc adm release mirror` writes an ICSP file to the current directory or wherever you specify 
+* Add logic to `oc adm release` to become aware of ICSP in cluster
+* Add logic to `oc adm release` to use the ICSP to complete extracts, mirroring, must-gather
+
+## User Stories
+
+A user runs:
+1. `oc adm must-gather`
+    * `oc` will look for ICSP from cluster (no flag necessary here, since this cmd is always run while connected to a cluster) and will use the mirror from ICSP to launch the must-gather pod.  If no ICSP found, will look in `-n openshift` for must-gather imagestream.
+2. `oc adm release extract --tools registry.example.com/repo/release:tag`
+    * `oc` will look for ICSP and if found, will extract from the ICSP mirror (aregistry/arepo/release:tag-toolsha) rather than from the image reference that the user will not
+    have access to if in disconnected environment.  If ICSP not found, will extract from the user-given registry.example.com/repo/tool:digest.  
+    If neither of these succeed, will proceed to use the image-reference from the user-given image.  The extract will succeed if user has access and permission to the
```

Comment:
Not sure which fallback this covers, you mentioned in the previous sentence that if ISCP is not found we'll extract from user-given image. So I'm a bit confused with this.

Reply:
So, if a user has mirrored an image from `quay.io/openshift-release-dev/ocp-release:tag` to `repo/release/ocp:tag`  then w/out this PR, when you extract from `repo/release/ocp:tag` oc adm release extract _only_ extracts from the original, `quay.io/openshift-release-dev/ocp-release:tag`.  

With this PR, result of a user running:
`$ oc adm release extract --tools repo/release/ocp:tag` will be:
1) Try to extract from ICSP file or ICSP from connected cluster
2) Try to extract from user-given release ( `repo/release/ocp:tag` ) 
3) If those fail, _then_ fall to extract from the image reference (the original image) .  This may or may not succeed, depending on if you have access to that registry thru pull-secret/reg credentials.